rust/connlib/tunnel/src/peer.rs

```diff
@@ -274,16 +294,37 @@ impl PacketTransform for PacketTransformClient {
         packet: &'a mut [u8],
     ) -> Result<(device_channel::Packet<'a>, IpAddr)> {
         let translations = self.translations.read();
-        let src = translations.get_by_right(addr).unwrap_or(addr);
+        let mut src = *translations.get_by_right(addr).unwrap_or(addr);
 
         let Some(mut pkt) = MutableIpPacket::new(packet) else {
             return Err(Error::BadPacket);
         };
 
-        pkt.set_src(*src);
+        let original_src = src;
+        if self
+            .upstream_dns
+            .load()
+            .as_ref()
+            .is_some_and(|upstream| **upstream == src)
+        {
+            if let Some(dgm) = pkt.as_udp() {
+                if let Ok(message) = domain::base::Message::from_slice(dgm.payload()) {
+                    if self
+                        .mangled_dns_ids
+                        .lock()
+                        .remove(&message.header().id())
+                        .is_some_and(|exp| exp.elapsed() < IDS_EXPIRE)
```

Comment:
So we check every packet, if it's UDP, then we also check if it's a valid DNS packet, if it is then we update the DNS tracking?

The `elapsed` could be removed here by storing the Instant when the track will expire instead of the instant when the track began, but since this is only called on DNS packets, it shouldn't affect performance much.

Reply:
> So we check every packet, if it's UDP, then we also check if it's a valid DNS packet, if it is then we update the DNS tracking?

and the destination ip is for the upstream DNS.

 > The `elapsed` could be removed here by storing the Instant when the track will expire instead of the instant when the track began, but since this is only called on DNS packets, it shouldn't affect performance much.

Even if we did that, we need to retrieve the current time and that's the syscall that could, in theory, be a performance hit. And the `std` time's module API is much nicer when checking elapsed time.

Reply:
Yeah. I think when I wrote this comment I thought it would be checking multiple timers but it does the hashmap lookup and only checks one, if it checks any. Sounds good.